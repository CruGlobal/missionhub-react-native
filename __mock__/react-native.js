import { NativeModules } from 'react-native';

// From https://github.com/facebook/react-native/issues/18279#issuecomment-374177940
// Will be made unnecessary once https://github.com/facebook/react-native/pull/18718 is released. It doesn't seem to be in 0.55.3
NativeModules.BlobModule = {
  ...NativeModules.BlobModule,
  addNetworkingHandler: jest.fn(),
};
