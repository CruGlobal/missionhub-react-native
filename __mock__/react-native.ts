jest.mock(
  'react-native/Libraries/Utilities/NativePlatformConstantsIOS',
  () => ({
    // @ts-ignore
    ...jest.requireActual(
      'react-native/Libraries/Utilities/NativePlatformConstantsIOS',
    ),
    getConstants: () => ({
      forceTouchAvailable: false,
      interfaceIdiom: 'en',
      isTesting: false,
      osVersion: 'ios',
      reactNativeVersion: { major: 60, minor: 1, patch: 0 },
      systemName: 'ios',
    }),
  }),
);
