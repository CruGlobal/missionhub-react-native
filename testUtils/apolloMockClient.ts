import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema, IMocks } from 'graphql-tools';
import { DocumentNode } from 'graphql';
import { buildClientSchema, IntrospectionQuery } from 'graphql/utilities';
import gql from 'fraql';
import { createMockerFromIntrospection } from 'fraql/mock';

import introspectionQuery from '../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks: IMocks = {}) => {
  const schema = buildClientSchema(
    (introspectionQuery as unknown) as IntrospectionQuery,
  );

  addMockFunctionsToSchema({
    schema,
    mocks: { ...globalMocks, ...mocks },
  });

  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new SchemaLink({ schema }),
  });
};

export const mockFragment = <TData>(
  fragmentDocument: DocumentNode,
  options?: {
    mocks: IMocks;
  },
): TData => {
  if (!fragmentDocument.loc) {
    throw 'Unable to get original gql query to create FraQL fragment.';
  }

  return createMockerFromIntrospection(introspectionQuery, {
    mocks: globalMocks,
  }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
