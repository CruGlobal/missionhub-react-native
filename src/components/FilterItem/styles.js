import { StyleSheet } from 'react-native';

import theme from '../../theme';

export default StyleSheet.create({
  row: {
    padding: 13,
    backgroundColor: theme.white,
    borderBottomWidth: theme.separatorHeight,
    borderBottomColor: theme.separatorColor,
  },
  switchRow: {
    paddingVertical: 7,
  },
  name: {
    flex: 1,
    fontSize: 14,
  },
  anyText: {
    width: theme.fullWidth / 2,
    textAlign: 'right',
    fontSize: 14,
    color: theme.grey1,
    paddingRight: 16,
  },
  anyIcon: {
    fontSize: 18,
    color: theme.grey1,
    paddingRight: 10,
  },
  checkIcon: {
    fontSize: 18,
    color: theme.primaryColor,
  },
  switch: {},
});
