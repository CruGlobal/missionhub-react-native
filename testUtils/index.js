import 'react-native';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Need to import for all translation screens to work properly
import '../src/i18n';

Enzyme.configure({ adapter: new Adapter() });

export const createMockStore = (state = {}) => {
  return {
    getState: jest.fn(() => (state)),
    dispatch: jest.fn((response) => Promise.resolve(response)),
    subscribe: jest.fn(),
  };
};

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (data) => {
  expect(renderer.create(data)).toMatchSnapshot();
};

export const renderShallow = (component, store) => {
  let renderedComponent = shallow(
    component,
    { context: { store: store } }
  );

  // If component has translation wrappers, dive deeper
  while (renderedComponent.is('Translate') || renderedComponent.is('I18n')) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (component, store) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};

export const mockFnWithParams = (obj, method, expectedReturn, ...expectedParams) => {
  return obj[method] = jest.fn(
    (...actualParams) => {
      expect(actualParams).toEqual(expectedParams);
      return expectedReturn;
    }
  );
};
