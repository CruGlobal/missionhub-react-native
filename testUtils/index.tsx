import React, { ReactElement, ReactNode } from 'react';
import 'react-native';
import { Provider } from 'react-redux';
import { Provider as ProviderLegacy } from 'react-redux-legacy';
import thunk from 'redux-thunk';
import configureStore, { MockStore } from 'redux-mock-store';
// eslint-disable-next-line import/named
import { NavigationParams, NavigationProvider } from 'react-navigation';
import { ApolloProvider } from '@apollo/react-hooks';
import { ReactTestRendererJSON } from 'react-test-renderer';
import { render, RenderAPI } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';
import Enzyme, { shallow as enzymeShallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { IMocks } from 'graphql-tools';

import { createApolloMockClient } from './apolloMockClient';
import { createNavigationProp } from './navigationHelpers';

Enzyme.configure({ adapter: new Adapter() });

export const createThunkStore = configureStore([thunk]);

interface RenderWithContextParams {
  initialState?: {};
  store?: MockStore;
  navParams?: NavigationParams;
  mocks?: IMocks;
  noWrappers?: boolean;
}

// Inspiration from https://github.com/kentcdodds/react-testing-library/blob/52575005579307bcfbe7fbe4ef4636147c03c6fb/examples/__tests__/react-redux.js#L69-L80
export function renderWithContext(
  component: ReactElement,
  {
    initialState,
    store = createThunkStore(initialState),
    navParams,
    mocks: mocks = {},
    noWrappers = false,
  }: RenderWithContextParams = {},
) {
  const mockApolloClient = createApolloMockClient(mocks);

  const navigation = createNavigationProp(navParams);

  let renderResult: RenderAPI;
  if (noWrappers) {
    renderResult = render(React.cloneElement(component, { navigation }));
  } else {
    // Warning: don't call any functions in here that return new instances on every call. All the props need to stay the same otherwise rerender won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
      <NavigationProvider value={navigation}>
        <ProviderLegacy store={store}>
          <Provider store={store}>
            <ApolloProvider client={mockApolloClient}>
              {children}
            </ApolloProvider>
          </Provider>
        </ProviderLegacy>
      </NavigationProvider>
    );

    renderResult = render(React.cloneElement(component, { navigation }), {
      wrapper,
    });
  }

  let storedSnapshot: ReactTestRendererJSON | null;
  return {
    ...renderResult,
    store,
    snapshot: () => expect(renderResult.toJSON()).toMatchSnapshot(),
    recordSnapshot: () => (storedSnapshot = renderResult.toJSON()),
    diffSnapshot: () => {
      if (!storedSnapshot) {
        throw new Error(
          'You must call recordSnapshot to store an initial snapshot before calling diffSnapshot',
        );
      }
      expect(
        snapshotDiff(storedSnapshot, renderResult.toJSON()),
      ).toMatchSnapshot();
    },
  };
}

// TODO: Remove all legacy rendering functions below

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (component: ReactElement) => {
  const { toJSON } = render(component);
  expect(toJSON()).toMatchSnapshot();
};

export const renderShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  let renderedComponent = enzymeShallow(
    <ProviderLegacy store={store}>{component}</ProviderLegacy>,
  ).dive();

  // If component has translation wrappers, dive deeper
  while (
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    ((renderedComponent.type() as any).displayName || '').startsWith(
      'withI18nextTranslation(',
    )
  ) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};
