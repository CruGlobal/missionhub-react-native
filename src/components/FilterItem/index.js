import React, { Component } from 'react';
import { Switch } from 'react-native';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';

import { Flex, Text, Touchable, Icon } from '../common';
import theme from '../../theme';

import styles from './styles';

@withTranslation('searchFilterRefine')
export default class FilterItem extends Component {
  handleSelect = () => {
    this.props.onSelect(this.props.item);
  };
  renderRight() {
    const { item, type, isSelected, t } = this.props;

    if (type === 'drilldown') {
      return (
        <Flex direction="row" align="center">
          <Text style={styles.anyText} numberOfLines={1}>
            {item.preview || t('any')}
          </Text>
          <Icon
            name="rightArrowIcon"
            type="MissionHub"
            style={styles.anyIcon}
          />
        </Flex>
      );
    }
    if (type === 'switch') {
      return (
        <Switch
          onValueChange={this.handleSelect}
          value={this.props.isSelected}
          onTintColor={theme.primaryColor}
        />
      );
    }
    if (type === 'single' && isSelected) {
      return (
        <Icon name="checkIcon" type="MissionHub" style={styles.checkIcon} />
      );
    }
    return null;
  }
  render() {
    const { item, type } = this.props;
    const content = (
      <Flex
        direction="row"
        align="center"
        style={[styles.row, type === 'switch' ? styles.switchRow : null]}
      >
        <Text style={styles.name}>{item.text}</Text>
        {this.renderRight()}
      </Flex>
    );

    if (type === 'drilldown' || type === 'single') {
      return (
        <Touchable highlight={true} onPress={this.handleSelect}>
          {content}
        </Touchable>
      );
    }
    return content;
  }
}

FilterItem.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string,
    text: PropTypes.string,
    preview: PropTypes.string,
  }).isRequired,
  onSelect: PropTypes.func.isRequired,
  type: PropTypes.oneOf(['drilldown', 'single', 'switch']).isRequired,
  isSelected: PropTypes.bool,
};
