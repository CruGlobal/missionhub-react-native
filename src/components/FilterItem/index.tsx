import React from 'react';
import { Switch } from 'react-native';
import { useTranslation } from 'react-i18next';

import { Flex, Text, Touchable, Icon } from '../common';
import theme from '../../theme';

import styles from './styles';

type FilterItemType = {
  id?: string;
  text?: string;
  preview?: string;
};

interface FilterItemProps {
  item: FilterItemType;
  onSelect: (item: FilterItemType) => void;
  type: 'drilldown' | 'single' | 'switch';
  isSelected?: boolean;
  testID?: string;
}

const FilterItem = ({ item, onSelect, type, isSelected }: FilterItemProps) => {
  const { t } = useTranslation('searchFilterRefine');

  const handleSelect = () => {
    onSelect(item);
  };

  const renderRight = () => {
    if (type === 'drilldown') {
      return (
        <Flex direction="row" align="center">
          <Text style={styles.anyText} numberOfLines={1}>
            {item.preview || t('any')}
          </Text>
          <Icon
            name="rightArrowIcon"
            type="MissionHub"
            style={styles.anyIcon}
          />
        </Flex>
      );
    }
    if (type === 'switch') {
      return (
        <Switch
          onValueChange={handleSelect}
          value={isSelected}
          trackColor={{ true: theme.primaryColor, false: '' }}
        />
      );
    }
    if (type === 'single' && isSelected) {
      return (
        <Icon name="checkIcon" type="MissionHub" style={styles.checkIcon} />
      );
    }
    return null;
  };

  const content = (
    <Flex
      direction="row"
      align="center"
      style={[styles.row, type === 'switch' ? styles.switchRow : null]}
    >
      <Text style={styles.name}>{item.text}</Text>
      {renderRight()}
    </Flex>
  );

  if (type === 'drilldown' || type === 'single') {
    return (
      <Touchable highlight={true} onPress={handleSelect}>
        {content}
      </Touchable>
    );
  }
  return content;
};

export default FilterItem;
