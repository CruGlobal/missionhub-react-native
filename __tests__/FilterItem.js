import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import FilterItem from '../src/components/FilterItem';
import { testSnapshot } from '../testUtils';

const mockItem = {
  id: '123',
  text: 'Test Step',
  preview: 'Preview',
};
const onSelect = jest.fn();

jest.mock('Switch');

it('renders single item correctly', () => {
  testSnapshot(
    <FilterItem item={mockItem} onSelect={onSelect} type="single" />,
  );
});

it('renders drilldown item correctly', () => {
  testSnapshot(
    <FilterItem item={mockItem} onSelect={onSelect} type="drilldown" />,
  );
});

it('renders switch item correctly', () => {
  testSnapshot(
    <FilterItem item={mockItem} onSelect={onSelect} type="switch" />,
  );
});

it('renders selected switch item correctly', () => {
  testSnapshot(
    <FilterItem
      item={mockItem}
      onSelect={onSelect}
      type="switch"
      isSelected={true}
    />,
  );
});
