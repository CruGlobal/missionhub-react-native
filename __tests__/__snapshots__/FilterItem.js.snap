// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders drilldown item correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits="button"
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      undefined,
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "borderBottomColor": "#E6E8EC",
            "borderBottomWidth": 1,
            "padding": 13,
          },
          null,
        ],
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#505256",
          },
          Object {
            "fontFamily": "SourceSansPro-Regular",
          },
          Object {
            "flex": 1,
            "fontSize": 14,
          },
        ]
      }
    >
      Test Step
    </Text>
    <View
      style={
        Array [
          Object {},
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#505256",
            },
            Object {
              "fontFamily": "SourceSansPro-Regular",
            },
            Object {
              "color": "#B2B0B2",
              "fontSize": 14,
              "paddingRight": 16,
              "textAlign": "right",
              "width": 375,
            },
          ]
        }
      >
        Preview
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Array [
              Object {
                "color": "#ffffff",
              },
              Object {
                "fontSize": 18,
              },
              Object {
                "color": "#B2B0B2",
                "fontSize": 18,
                "paddingRight": 10,
              },
            ],
            Object {
              "fontFamily": "icomoon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        юда
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders selected switch item correctly 1`] = `
<View
  style={
    Array [
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "borderBottomColor": "#E6E8EC",
          "borderBottomWidth": 1,
          "padding": 13,
        },
        Object {
          "paddingVertical": 7,
        },
      ],
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#505256",
        },
        Object {
          "fontFamily": "SourceSansPro-Regular",
        },
        Object {
          "flex": 1,
          "fontSize": 14,
        },
      ]
    }
  >
    Test Step
  </Text>
</View>
`;

exports[`renders single item correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits="button"
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      undefined,
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "borderBottomColor": "#E6E8EC",
            "borderBottomWidth": 1,
            "padding": 13,
          },
          null,
        ],
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#505256",
          },
          Object {
            "fontFamily": "SourceSansPro-Regular",
          },
          Object {
            "flex": 1,
            "fontSize": 14,
          },
        ]
      }
    >
      Test Step
    </Text>
  </View>
</View>
`;

exports[`renders switch item correctly 1`] = `
<View
  style={
    Array [
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "borderBottomColor": "#E6E8EC",
          "borderBottomWidth": 1,
          "padding": 13,
        },
        Object {
          "paddingVertical": 7,
        },
      ],
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#505256",
        },
        Object {
          "fontFamily": "SourceSansPro-Regular",
        },
        Object {
          "flex": 1,
          "fontSize": 14,
        },
      ]
    }
  >
    Test Step
  </Text>
</View>
`;
